api/crypto/frame_crypto_transformer.cc

```diff
@@ -470,35 +517,88 @@ void FrameCryptorTransformer::decryptFrame(
     encrypted_payload[i - unencrypted_bytes] = date_in[i];
   }
   std::vector<uint8_t> buffer;
-  if (AesEncryptDecrypt(EncryptOrDecrypt::kDecrypt, algorithm_, aes_key, iv,
-                        frameHeader, encrypted_payload, &buffer) == Success) {
-    rtc::Buffer payload(buffer.data(), buffer.size());
-    rtc::Buffer data_out;
-    data_out.AppendData(frameHeader);
-    data_out.AppendData(payload);
-    frame->SetData(data_out);
 
-    RTC_LOG(LS_INFO) << "FrameCryptorTransformer::decryptFrame() ivLength="
-                     << static_cast<int>(ivLength) << " unencrypted_bytes="
-                     << static_cast<int>(unencrypted_bytes)
-                     << " keyIndex=" << static_cast<int>(key_index_)
-                     << " aesKey=" << to_hex(aes_key.data(), aes_key.size())
-                     << " iv=" << to_hex(iv.data(), iv.size());
+  int ratchet_count = 0;
+  auto initialKey = key_set->material;
+  bool decryption_success = false;
+  if (AesEncryptDecrypt(EncryptOrDecrypt::kDecrypt, algorithm_,
+                        key_set->encryption_key, iv, frameHeader,
+                        encrypted_payload, &buffer) == Success) {
+    decryption_success = true;
+  } else {
+    RTC_LOG(LS_ERROR) << "FrameCryptorTransformer::decryptFrame() failed";
 
-    if (last_dec_error_ != FrameCryptionError::kOk) {
-      last_dec_error_ = FrameCryptionError::kOk;
-      if (observer_)
-        observer_->OnFrameCryptionError(participant_id_, last_dec_error_);
+    if (key_handler->options().ratchet_window_size > 0) {
+      while (ratchet_count < key_handler->options().ratchet_window_size) {
+        ratchet_count++;
+
+        RTC_LOG(LS_INFO) << "ratcheting key attempt " << ratchet_count << " of "
+                         << key_handler->options().ratchet_window_size;
+
+        key_handler->RatchetKey(key_index);
```

Comment:
what I did in the JS implementation now is that 
`RatchetKey` returns the key directly and there's a parameter to specifiy whether or not to automatically set the key after ratcheting. 
This allows us to say `give me the next ratcheted key, but only set it on the key manager as the correct one, if we know that decryption was successful`.

Reply:
@lukasIO  updated https://github.com/webrtc-sdk/webrtc/pull/66/commits/faa419cee00cdab7780ee14532f531d0421ce167